Nix build: gomodcache hash drift breaks reproducibility

## Problem

The `effect-tsgo-workspace-gomodcache` fixed-output derivation in `flake.nix` uses a hardcoded `workspaceModuleCacheHash` that breaks when the Go module proxy serves different bytes for the same module versions:

```
error: hash mismatch in fixed-output derivation '.../effect-tsgo-workspace-gomodcache.drv':
         specified: sha256-LYZFR4Km+fK1ZnSp7yUB8Tc+h+kOboZolU6uDnEASw0=
            got:    sha256-RkMqS8w2OaGTdI3kdLW8yP/skuryULhqdHNNItB0lO0=
```

This happens at commit bd1258197942f077785183d8246ad9429a6902ea — the code hasn't changed, but the Go module proxy now returns different content (retracted/re-published transitive dep, proxy cache refresh, etc.).

This is a known fragility of Go fixed-output derivations that use `go mod download all` — the output isn't guaranteed to be bit-for-bit stable across time.

## Impact

Any downstream consumer pinned to this commit (via Nix flake input) can no longer build `effect-tsgo` without manually updating the hash. CI pipelines break with no code change.

## Suggestion

Consider a more robust approach, e.g.:

- **`buildGoModule` with `vendorHash`** — nixpkgs' `buildGoModule` already handles this pattern and has better caching semantics. Commit 831fca0 moved to `buildGoModule` but the workspace gomodcache derivation still exists alongside it.
- **Vendor the Go modules** (`go mod vendor`) — checked into the repo or as a separate derivation, this eliminates proxy non-determinism entirely.
- **Use `gomod2nix`** — generates a locked JSON manifest of exact module hashes, immune to proxy drift.

Any of these would avoid the brittle `go mod download all` + hardcoded hash approach that breaks when upstream module proxies aren't perfectly reproducible.

## Additional: nil pointer crash on latest main

Separately, the latest commit (5b09a5ec) panics when running `tsgo --build` on a project using Effect:

```
panic: runtime error: invalid memory address or nil pointer dereference
goroutine 5448 [running]:
...
github.com/effect-ts/effect-typescript-go/internal/typeparser.EffectType(...)
  effect_type.go:32
```

The previously pinned revision (bd125819) works fine on the same codebase.

---

*Filed on behalf of @schickling*

--- a/_tools/update-flake-vendor-hash.sh
+++ b/_tools/update-flake-vendor-hash.sh
@@ -42,19 +42,24 @@ update_input_commit() {
   local suffix="${4:-}"
 
   local current
-  current="$(perl -ne "
-    if (/\"github:${repo_url}\\/([a-f0-9]+)${suffix}\"/) {
-      print \"\$1\\n\";
+  current="$(REPO_URL="$repo_url" SUFFIX="$suffix" perl -ne '
+    my $pat = quotemeta("github:$ENV{REPO_URL}/");
+    my $suf = quotemeta($ENV{SUFFIX});
+    if (m{${pat}([a-f0-9]+)${suf}}) {
+      print "$1\n";
       exit 0;
     }
-  " "$flake_file")"
+  ' "$flake_file")"
 
   if [[ "$current" == "$new_commit" ]]; then
     echo "$input_name: already at $new_commit"
   else
-    local escaped_suffix
-    escaped_suffix="$(printf '%s' "$suffix" | sed 's/[?]/\\?/g')"
-    perl -pi -e "s|github:${repo_url}/[a-f0-9]+${escaped_suffix}|github:${repo_url}/${new_commit}${suffix}|" "$flake_file"
+    REPO_URL="$repo_url" OLD_COMMIT="$current" NEW_COMMIT="$new_commit" SUFFIX="$suffix" \
+      perl -pi -e '
+        my $old = quotemeta("github:$ENV{REPO_URL}/$ENV{OLD_COMMIT}$ENV{SUFFIX}");
+        my $new = "github:$ENV{REPO_URL}/$ENV{NEW_COMMIT}$ENV{SUFFIX}";
+        s/$old/$new/;
+      ' "$flake_file"
     echo "$input_name: updated ${current:-unknown} -> $new_commit"
   fi
 }

--- a/flake.nix
+++ b/flake.nix
@@ -36,19 +36,15 @@
       ];
       /*
        Go module vendor hash for buildGoModule (proxyVendor mode).
-       Uses `go mod download` under the hood — the download cache uses `!`
-       escaping for uppercase letters in module paths, making it deterministic
-       across case-sensitive (Linux) and case-insensitive (macOS APFS)
-       filesystems.
+       proxyVendor is required because this project has deps with
+       mixed-case module paths (Microsoft/go-winio vs microsoft/
+       typescript-go) — `go work vendor` produces different directory
+       layouts on case-sensitive (Linux) vs case-insensitive (macOS)
+       filesystems. The download cache uses `!` escaping for uppercase
+       letters, making it deterministic across both.
 
-       Refresh workflow:
-       1. Run `./_tools/update-flake-vendor-hash.sh`.
-       2. Commit the resulting `flake.nix` update if the script changed it.
-
-       Manual fallback:
-       1. Temporarily set this value to `lib.fakeHash`.
-       2. Run `nix build .#effect-tsgo --no-write-lock-file`.
-       3. Copy the reported `got: sha256-...` value back here.
+       Refresh: ./_tools/update-flake-vendor-hash.sh
+       Manual:  set to lib.fakeHash, build, copy the reported hash.
       */
       vendorHash = "sha256-15dQYcuOPxrnIXWV8UVVSqDqMyRJJmQqVEF7o3wgxz4=";
       forAllSystems =

--- a/internal/typeparser/effect_type.go
+++ b/internal/typeparser/effect_type.go
@@ -13,6 +13,9 @@ import (
 // The detection strategy is chosen based on the detected Effect version:
 // v4 uses direct symbol lookup, v3/unknown uses property iteration.
 func EffectType(c *checker.Checker, t *checker.Type, atLocation *ast.Node) *Effect {
+	if c == nil || t == nil {
+		return nil
+	}
 	version := DetectEffectVersion(c)
 	if version == EffectMajorV4 {
 		// Direct property access using the known Effect v4 type ID
@@ -128,6 +131,9 @@ func StrictIsEffectType(c *checker.Checker, t *checker.Type, atLocation *ast.Nod
 // For v4, this is a quick check for the "~effect/Effect" property.
 // For v3/unknown, this defers to IsEffectType since there is no single property shortcut.
 func HasEffectTypeId(c *checker.Checker, t *checker.Type, atLocation *ast.Node) bool {
+	if c == nil || t == nil {
+		return false
+	}
 	version := DetectEffectVersion(c)
 	if version == EffectMajorV4 {
 		return GetPropertyOfTypeByName(c, t, EffectTypeId) != nil

--- a/internal/typeparser/expected_and_real_type_test.go
+++ b/internal/typeparser/expected_and_real_type_test.go
@@ -287,6 +287,57 @@ func TestExpectedAndRealTypes_NilInputs(t *testing.T) {
 	}
 }
 
+func TestEffectType_NilInputs(t *testing.T) {
+	t.Parallel()
+
+	source := `const x: number = 42`
+	c, sf, done := compileAndGetCheckerAndSourceFile(t, source)
+	defer done()
+
+	// nil type must not panic
+	if result := typeparser.EffectType(c, nil, sf.AsNode()); result != nil {
+		t.Error("expected nil for nil type")
+	}
+	// nil checker must not panic
+	if result := typeparser.EffectType(nil, nil, sf.AsNode()); result != nil {
+		t.Error("expected nil for nil checker")
+	}
+	// HasEffectTypeId with nil type must not panic
+	if typeparser.HasEffectTypeId(c, nil, sf.AsNode()) {
+		t.Error("expected false for nil type")
+	}
+}
+
+func TestLayerType_NilInputs(t *testing.T) {
+	t.Parallel()
+
+	source := `const x: number = 42`
+	c, sf, done := compileAndGetCheckerAndSourceFile(t, source)
+	defer done()
+
+	if result := typeparser.LayerType(c, nil, sf.AsNode()); result != nil {
+		t.Error("expected nil for nil type")
+	}
+	if result := typeparser.LayerType(nil, nil, sf.AsNode()); result != nil {
+		t.Error("expected nil for nil checker")
+	}
+}
+
+func TestContextTag_NilInputs(t *testing.T) {
+	t.Parallel()
+
+	source := `const x: number = 42`
+	c, sf, done := compileAndGetCheckerAndSourceFile(t, source)
+	defer done()
+
+	if result := typeparser.ContextTag(c, nil, sf.AsNode()); result != nil {
+		t.Error("expected nil for nil type")
+	}
+	if result := typeparser.ContextTag(nil, nil, sf.AsNode()); result != nil {
+		t.Error("expected nil for nil checker")
+	}
+}
+
 func TestExpectedAndRealTypes_EmptyFile(t *testing.T) {
 	t.Parallel()
 

--- a/internal/typeparser/layer_type.go
+++ b/internal/typeparser/layer_type.go
@@ -33,6 +33,9 @@ func parseLayerVarianceStruct(c *checker.Checker, t *checker.Type, atLocation *a
 // The detection strategy is chosen based on the detected Effect version:
 // v4 uses direct symbol lookup, v3/unknown uses property iteration.
 func LayerType(c *checker.Checker, t *checker.Type, atLocation *ast.Node) *Layer {
+	if c == nil || t == nil {
+		return nil
+	}
 	version := DetectEffectVersion(c)
 	if version == EffectMajorV4 {
 		// Direct property access using the known Layer v4 type ID

--- a/internal/typeparser/service_type.go
+++ b/internal/typeparser/service_type.go
@@ -47,6 +47,9 @@ func IsServiceType(c *checker.Checker, t *checker.Type, atLocation *ast.Node) bo
 // For V3/unknown, this iterates properties looking for a service variance struct,
 // following the same pattern as LayerType() and EffectType().
 func ContextTag(c *checker.Checker, t *checker.Type, atLocation *ast.Node) *Service {
+	if c == nil || t == nil {
+		return nil
+	}
 	version := DetectEffectVersion(c)
 	if version == EffectMajorV4 {
 		return ServiceType(c, t, atLocation)

